iwyu removes some headers incorrectly

I have these three files:

FontMgr.h:
```c++
class IFontMgr {
 public:
  void getFont();
};
```

main.h:
```c++
#include <memory>

class IFontMgr;

class FontRenderer {
  void render() const;

 private:
  std::shared_ptr<IFontMgr> fontMgr_;
};
```

main.cpp:
```c++
#include "main.h"
#include "FontMgr.h"

void FontRenderer::render() const {
  fontMgr_->getFont();
}
```

Running iwyu(main branch & clang_16 branch):
```bash
$ include-what-you-use -stdlib=libc++ main.cpp
```

I get:
```
main.h should add these lines:
#include <__memory/shared_ptr.h>  // for shared_ptr

main.h should remove these lines:
- #include <memory>  // lines 1-1

The full include-list for main.h:
#include <__memory/shared_ptr.h>  // for shared_ptr
class IFontMgr;  // lines 3-3
---

main.cpp should add these lines:

main.cpp should remove these lines:
- #include "FontMgr.h"  // lines 2-2

The full include-list for main.cpp:
#include "main.h"
---
```

`FontMgr.h` is removed by iwyu, and `main.cpp` cannot be compiled by clang, I get errors like:
```
error: member access into incomplete type 'element_type' (aka 'IFontMgr')
```

This happens not only when calling `shared_ptr<IFontMgr>->method()`, but also `optional<IFontMgr>->method()`.

I increased log verbose and got:
```
--- Calculating IWYU violations for main.h ---
Ignoring fwd-decl use of FontRenderer (main.h:5:7): dfn is present: main.h:5:7
Ignoring fwd-decl use of FontRenderer (main.h:5:7): dfn is present: main.h:5:7
Ignoring use of FontRenderer (main.h:5:7): definition is present: main.h:5:7
Ignoring use of FontRenderer (main.h:5:7): definition is present: main.h:5:7
Ignoring use of std::shared_ptr::~shared_ptr<_Tp> (main.h:5:7): member of class
Ignoring use of std::shared_ptr::shared_ptr<_Tp> (main.h:5:7): member of class
Mapped /usr/lib/llvm-16/include/c++/v1/__memory/shared_ptr.h to <__memory/shared_ptr.h> for std::shared_ptr (only candidate)
Noting fwd-decl use of IFontMgr (main.h:9:19) is declared at main.h:3:7
Ignoring fwd-decl use of IFontMgr (main.h:9:19): have earlier fwd-decl at main.h:3:7
main.h:9:8: warning: std::shared_ptr is defined in <__memory/shared_ptr.h>, which isn't directly #included.

main.h should add these lines:
#include <__memory/shared_ptr.h>  // for shared_ptr

main.h should remove these lines:
- #include <memory>  // lines 1-1

The full include-list for main.h:
#include <__memory/shared_ptr.h>  // for shared_ptr
class IFontMgr;  // lines 3-3
---
--- Calculating IWYU violations for main.cpp ---
Ignoring use of std::shared_ptr::operator-> (main.cpp:5:11): member of class
Ignoring use of std::shared_ptr::operator-> (main.cpp:5:11): member of class
Ignoring use of std::shared_ptr::operator-> (main.cpp:5:11): member of class
Ignoring use of IFontMgr::getFont (main.cpp:5:3): member of class
Marked use of include-file  (from "main.h") at <invalid loc>
Mapped main.h to "main.h" for FontRenderer (only candidate)
Mapped /usr/lib/llvm-16/include/c++/v1/__memory/shared_ptr.h to <__memory/shared_ptr.h> for std::shared_ptr (only candidate)
Mapped main.h to "main.h" for FontRenderer (only candidate)
Ignoring full use of FontRenderer (main.cpp:4:6): #including dfn from "main.h"
Ignoring full use of FontRenderer (main.cpp:5:3): #including dfn from "main.h"
Ignoring full use of  (<invalid loc>): #including dfn from "main.h"
main.cpp:5:3: warning: std::shared_ptr is defined in <__memory/shared_ptr.h>, which isn't directly #included.

main.cpp should add these lines:

main.cpp should remove these lines:
- #include "FontMgr.h"  // lines 2-2

The full include-list for main.cpp:
#include "main.h"
---
```

Seems that the use of FontMgr is ignored because of "member of class".

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3090,6 +3090,9 @@ class InstantiatedTemplateVisitor
     if (CanIgnoreCurrentASTNode())
       return true;
 
+    if (type->isTypeAlias())
+      return TraverseType(type->getAliasedType());
+
     // Skip the template traversal if this occurrence of the template name is
     // just a class qualifier for an out of line method, as opposed to an object
     // instantiation, where the templated code would need to be inspected.
@@ -4219,7 +4222,7 @@ class IwyuAstConsumer
 
     // If we're not in a forward-declare context, use of a template
     // specialization requires having the full type information.
-    if (!CanForwardDeclareType(current_ast_node())) {
+    if (!CanForwardDeclareType(current_ast_node()) || type->isTypeAlias()) {
       const TemplateInstantiationData data = GetTplInstData(type);
       instantiated_template_visitor_.ScanInstantiatedType(
           current_ast_node(), data.resugar_map, data.provided_types);

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -1447,8 +1447,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
-  set<const Type*> GetProvidedTypesForTypedef(
-      const TypedefNameDecl* decl) const {
+  set<const Type*> GetProvidedTypesForAlias(const TypedefNameDecl* decl) const {
     const Type* underlying_type = decl->getUnderlyingType().getTypePtr();
     // If the underlying type is itself a typedef, we recurse.
     if (const auto* underlying_typedef =
@@ -1457,7 +1456,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         // TODO(csilvers): if one of the intermediate typedefs
         // #includes the necessary definition of the 'final'
         // underlying type, do we want to return it here?
-        return GetProvidedTypesForTypedef(underlying_typedef_decl);
+        return GetProvidedTypesForAlias(underlying_typedef_decl);
       }
     }
     // TODO(bolshakov): handle underlying alias templates.
@@ -2672,7 +2671,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       if (!ast_node->ParentIsA<TypedefNameDecl>()) {
         const TypedefNameDecl* typedef_decl = typedef_type->getDecl();
         const set<const Type*>& provided_with_typedef =
-            GetProvidedTypesForTypedef(typedef_decl);
+            GetProvidedTypesForAlias(typedef_decl);
         InsertAllInto(provided_with_typedef, &blocked_types);
         VERRS(6) << "User, not author, of typedef "
                  << typedef_decl->getQualifiedNameAsString()
@@ -2690,7 +2689,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         const NamedDecl* decl = TypeToDeclAsWritten(template_spec_type);
         if (const auto* al_tpl_decl = dyn_cast<TypeAliasTemplateDecl>(decl)) {
           const TypeAliasDecl* al_decl = al_tpl_decl->getTemplatedDecl();
-          InsertAllInto(GetProvidedTypesForTypedef(al_decl), &blocked_types);
+          InsertAllInto(GetProvidedTypesForAlias(al_decl), &blocked_types);
           const Type* type = template_spec_type->getAliasedType().getTypePtr();
           ReportTypeUseInternal(used_loc, type, blocked_types, deref_kind);
         }
@@ -4377,7 +4376,7 @@ class IwyuAstConsumer
     const Type* desugared_until_typedef = Desugar(type);
     if (const auto* typedef_type =
             dyn_cast_or_null<TypedefType>(desugared_until_typedef)) {
-      return GetProvidedTypesForTypedef(typedef_type->getDecl());
+      return GetProvidedTypesForAlias(typedef_type->getDecl());
     }
     // TODO(bolshakov): handle alias templates.
     return set<const Type*>();

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -1447,8 +1447,8 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
-  set<const Type*> GetProvidedTypesForAlias(const TypedefNameDecl* decl) const {
-    const Type* underlying_type = decl->getUnderlyingType().getTypePtr();
+  set<const Type*> GetProvidedTypesForAlias(const Type* underlying_type,
+                                            SourceLocation decl_loc) const {
     // If the underlying type is itself a typedef, we recurse.
     if (const auto* underlying_typedef =
             underlying_type->getAs<TypedefType>()) {
@@ -1456,12 +1456,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         // TODO(csilvers): if one of the intermediate typedefs
         // #includes the necessary definition of the 'final'
         // underlying type, do we want to return it here?
-        return GetProvidedTypesForAlias(underlying_typedef_decl);
+        return GetProvidedTypesForAlias(
+            underlying_typedef_decl->getUnderlyingType().getTypePtr(),
+            GetLocation(underlying_typedef_decl));
       }
     }
     // TODO(bolshakov): handle underlying alias templates.
 
-    return GetProvidedTypes(underlying_type, GetLocation(decl));
+    return GetProvidedTypes(underlying_type, decl_loc);
   }
 
   // ast_node is the node for the autocast CastExpr.  We use it to get
@@ -2670,15 +2672,15 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       const ASTNode* ast_node = MostElaboratedAncestor(current_ast_node());
       if (!ast_node->ParentIsA<TypedefNameDecl>()) {
         const TypedefNameDecl* typedef_decl = typedef_type->getDecl();
+        const Type* type = typedef_decl->getUnderlyingType().getTypePtr();
         const set<const Type*>& provided_with_typedef =
-            GetProvidedTypesForAlias(typedef_decl);
+            GetProvidedTypesForAlias(type, GetLocation(typedef_decl));
         InsertAllInto(provided_with_typedef, &blocked_types);
         VERRS(6) << "User, not author, of typedef "
                  << typedef_decl->getQualifiedNameAsString()
                  << " owns the underlying type:\n";
         // If any of the used types are themselves typedefs, this will
         // result in a recursive expansion.
-        const Type* type = typedef_decl->getUnderlyingType().getTypePtr();
         ReportTypeUseInternal(used_loc, type, blocked_types, deref_kind);
       }
       return;
@@ -2689,8 +2691,9 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         const NamedDecl* decl = TypeToDeclAsWritten(template_spec_type);
         if (const auto* al_tpl_decl = dyn_cast<TypeAliasTemplateDecl>(decl)) {
           const TypeAliasDecl* al_decl = al_tpl_decl->getTemplatedDecl();
-          InsertAllInto(GetProvidedTypesForAlias(al_decl), &blocked_types);
           const Type* type = template_spec_type->getAliasedType().getTypePtr();
+          InsertAllInto(GetProvidedTypesForAlias(type, GetLocation(al_decl)),
+                        &blocked_types);
           ReportTypeUseInternal(used_loc, type, blocked_types, deref_kind);
         }
         return;
@@ -4376,7 +4379,9 @@ class IwyuAstConsumer
     const Type* desugared_until_typedef = Desugar(type);
     if (const auto* typedef_type =
             dyn_cast_or_null<TypedefType>(desugared_until_typedef)) {
-      return GetProvidedTypesForAlias(typedef_type->getDecl());
+      const TypedefNameDecl* decl = typedef_type->getDecl();
+      return GetProvidedTypesForAlias(decl->getUnderlyingType().getTypePtr(),
+                                      GetLocation(decl));
     }
     // TODO(bolshakov): handle alias templates.
     return set<const Type*>();

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -249,6 +249,7 @@ using clang::TemplateTypeParmDecl;
 using clang::TranslationUnitDecl;
 using clang::Type;
 using clang::TypeAliasDecl;
+using clang::TypeAliasTemplateDecl;
 using clang::TypeLoc;
 using clang::TypeSourceInfo;
 using clang::TypeTrait;
@@ -1459,6 +1460,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         return GetProvidedTypesForTypedef(underlying_typedef_decl);
       }
     }
+    // TODO(bolshakov): handle underlying alias templates.
 
     return GetProvidedTypes(underlying_type, GetLocation(decl));
   }
@@ -2423,7 +2425,10 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       ReportDeclForwardDeclareUse(CurrentLoc(), decl);
       current_ast_node()->set_in_forward_declare_context(true);
     } else {
-      ReportDeclUse(CurrentLoc(), decl);
+      if (type->isTypeAlias())
+        ReportWrittenTypeAlias(type);
+      else
+        ReportDeclUse(CurrentLoc(), decl);
     }
 
     return true;
@@ -2432,18 +2437,8 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   bool VisitTypedefType(TypedefType* type) {
     if (CanIgnoreCurrentASTNode())
       return true;
-    if (!CanForwardDeclareType(current_ast_node())) {
-      // When the alias is from template (like 'Type' in
-      // 'Tpl<Providing>::Type'), the type may be provided with template
-      // argument and should not be reported. 'ReportTypeUse' performs that
-      // analysis when passing 'ElaboratedType' into it.
-      if (const auto* elaborated =
-              current_ast_node()->template GetParentAs<ElaboratedType>()) {
-        ReportTypeUse(CurrentLoc(), elaborated, DerefKind::None);
-      } else {
-        ReportTypeUse(CurrentLoc(), type, DerefKind::None);
-      }
-    }
+    if (!CanForwardDeclareType(current_ast_node()))
+      ReportWrittenTypeAlias(type);
     return true;
   }
 
@@ -2689,6 +2684,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       }
       return;
     }
+    if (const auto* template_spec_type =
+            type->getAs<TemplateSpecializationType>()) {
+      if (template_spec_type->isTypeAlias()) {
+        const NamedDecl* decl = TypeToDeclAsWritten(template_spec_type);
+        if (const auto* al_tpl_decl = dyn_cast<TypeAliasTemplateDecl>(decl)) {
+          const TypeAliasDecl* al_decl = al_tpl_decl->getTemplatedDecl();
+          InsertAllInto(GetProvidedTypesForTypedef(al_decl), &blocked_types);
+          const Type* type = template_spec_type->getAliasedType().getTypePtr();
+          ReportTypeUseInternal(used_loc, type, blocked_types, deref_kind);
+        }
+        return;
+      }
+    }
 
     // Map private types like __normal_iterator to their public counterpart.
     type = MapPrivateTypeToPublicType(type);
@@ -2708,6 +2716,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       decl = GetDefinitionAsWritten(decl);
       VERRS(6) << "(For type " << PrintableType(type) << "):\n";
       IwyuBaseAstVisitor<Derived>::ReportDeclUse(used_loc, decl);
+    }
+  }
+
+  void ReportWrittenTypeAlias(const Type* type) {
+    // When the alias is from template (like 'Type' in
+    // 'Tpl<Providing>::Type'), the type may be provided with template
+    // argument and should not be reported. 'ReportTypeUse' performs that
+    // analysis when passing 'ElaboratedType' into it.
+    if (const auto* elaborated =
+            current_ast_node()->template GetParentAs<ElaboratedType>()) {
+      ReportTypeUse(CurrentLoc(), elaborated, DerefKind::None);
+    } else {
+      ReportTypeUse(CurrentLoc(), type, DerefKind::None);
     }
   }
 
@@ -4207,7 +4228,7 @@ class IwyuAstConsumer
     // Don't call ReportTypeUse here, because scanning of template instantiation
     // internals should be avoided: it is allowed not to provide some of
     // template args.
-    if (is_provided)
+    if (is_provided || type->isTypeAlias())
       ReportDeclUse(CurrentLoc(), TypeToDeclAsWritten(type), comment);
 
     return Base::VisitTemplateSpecializationType(type);
@@ -4358,6 +4379,7 @@ class IwyuAstConsumer
             dyn_cast_or_null<TypedefType>(desugared_until_typedef)) {
       return GetProvidedTypesForTypedef(typedef_type->getDecl());
     }
+    // TODO(bolshakov): handle alias templates.
     return set<const Type*>();
   }
 

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -4332,13 +4332,19 @@ class IwyuAstConsumer
 
   set<const Type*> GetProvidedByTplArg(const Type* type) {
     set<const Type*> res;
-    if (const auto* elaborated = dyn_cast_or_null<ElaboratedType>(type)) {
+    if (!type)
+      return res;
+    if (const auto* elaborated = dyn_cast<ElaboratedType>(type)) {
       const NestedNameSpecifier* nns = elaborated->getQualifier();
       while (nns) {
         if (const Type* host = nns->getAsType())
           InsertAllInto(GetTplInstData(host).provided_types, &res);
         nns = nns->getPrefix();
       }
+    }
+    if (const auto* tpl_spec = type->getAs<TemplateSpecializationType>()) {
+      if (tpl_spec->isTypeAlias())
+        InsertAllInto(GetTplInstData(tpl_spec).provided_types, &res);
     }
     return res;
   }

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3456,12 +3456,6 @@ class InstantiatedTemplateVisitor
     if (current_ast_node()->in_forward_declare_context())
       return true;
 
-    while (type->isTypeAlias()) {
-      type = type->getAliasedType()->getAs<TemplateSpecializationType>();
-      if (!type)
-        return true;
-    }
-
     // If we're a dependent type, we only try to be analyzed if we're
     // in the precomputed list -- in general, the only thing clang
     // tells us about dependent types is their name (which is all we

--- a/iwyu_ast_util.cc
+++ b/iwyu_ast_util.cc
@@ -563,6 +563,17 @@ static const Type* GetTemplateArgAsType(const TemplateArgument& tpl_arg) {
   return nullptr;
 }
 
+static bool IsSubstTemplateTypeParmType(const Type* type) {
+  const Type* prev_type = type;
+  do {
+    if (isa<SubstTemplateTypeParmType>(type))
+      return true;
+    prev_type = type;
+    type = type->getLocallyUnqualifiedSingleStepDesugaredType().getTypePtr();
+  } while (type != prev_type);
+  return false;
+}
+
 // These utilities figure out the template arguments and other type components
 // that are specified in various contexts: TemplateSpecializationType (for
 // template classes), FunctionDecl (for template functions), and type aliases.
@@ -667,7 +678,7 @@ class TypeEnumeratorWithoutSubstituted
   bool TraverseTypeHelper(QualType qual_type) {
     CHECK_(!qual_type.isNull());
     const Type* type = qual_type.getTypePtr();
-    if (type->getAs<SubstTemplateTypeParmType>())
+    if (IsSubstTemplateTypeParmType(type))
       return true;
 
     seen_types_.insert(GetCanonicalType(type));

--- a/tests/cxx/alias_template.cc
+++ b/tests/cxx/alias_template.cc
@@ -13,6 +13,17 @@
 
 #include "tests/cxx/direct.h"
 
+template <typename T>
+using Identity = T;
+
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+Identity<IndirectClass> ic;
+// IWYU: IndirectClass is...*indirect.h
+constexpr auto s = sizeof(ic);
+// IWYU: IndirectClass needs a declaration
+Identity<IndirectClass>* pic = nullptr;
+
 template<class T> struct FullUseTemplateArgInSizeof {
   char argument[sizeof(T)];
 };

--- a/tests/cxx/alias_template.cc
+++ b/tests/cxx/alias_template.cc
@@ -70,6 +70,26 @@ using AliasNested2 = FullUseTemplateArgInSizeof<FullUseTemplateArgInSizeof<T>>;
 // IWYU: IndirectClass is...*indirect.h
 AliasNested2<IndirectClass> aliasNested2;
 
+template <typename T>
+using UsingArgInternals = decltype(T::a);
+
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+UsingArgInternals<IndirectClass> aliased_int;
+// Full type is needed even in fwd-decl context.
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+UsingArgInternals<IndirectClass>* p_int = nullptr;
+
+template <typename T>
+struct TplWithUsingArgInternals {
+  UsingArgInternals<T>* p = nullptr;
+};
+
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+TplWithUsingArgInternals<IndirectClass> twuai;
+
 /**** IWYU_SUMMARY
 
 tests/cxx/alias_template.cc should add these lines:

--- a/tests/cxx/alias_template.cc
+++ b/tests/cxx/alias_template.cc
@@ -24,6 +24,11 @@ constexpr auto s = sizeof(ic);
 // IWYU: IndirectClass needs a declaration
 Identity<IndirectClass>* pic = nullptr;
 
+// IWYU: IndirectClass is...*indirect.h
+using Providing = IndirectClass;
+
+Identity<Providing> type_is_provided_by_arg;
+
 template<class T> struct FullUseTemplateArgInSizeof {
   char argument[sizeof(T)];
 };

--- a/tests/cxx/iwyu_stricter_than_cpp-type_alias.h
+++ b/tests/cxx/iwyu_stricter_than_cpp-type_alias.h
@@ -91,6 +91,22 @@ using IndirectStruct3NonProvidingAl = IndirectStruct3;
 struct IndirectStruct4;
 using IndirectStruct4NonProvidingAl = IndirectStruct4;
 
+// --- Alias templates.
+
+template <typename>
+// IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h
+using DoesNotForwardDeclareAlTpl = IndirectStruct1;
+
+// struct DirectStruct1;  <-- Already present above.
+template <typename>
+using IncludesAlTpl = DirectStruct1;
+
+template <typename>
+using DoesNotForwardDeclareAndIncludesAlTpl = DirectStruct2;
+
+// struct IndirectStruct2;  <-- Already present above.
+template <typename>
+using DoesEverythingRightAlTpl = IndirectStruct2;
 
 /**** IWYU_SUMMARY
 

--- a/tests/cxx/iwyu_stricter_than_cpp-type_alias.h
+++ b/tests/cxx/iwyu_stricter_than_cpp-type_alias.h
@@ -108,6 +108,21 @@ using DoesNotForwardDeclareAndIncludesAlTpl = DirectStruct2;
 template <typename>
 using DoesEverythingRightAlTpl = IndirectStruct2;
 
+// --- Alias templates aliasing template specializations.
+
+template <typename T>
+using TemplateProvidedArgumentUsed = TplDirectStruct7<T, IndirectStruct2>;
+
+template <typename T>
+using TemplateNotProvidedArgumentUsed = TplIndirectStruct3<T, IndirectStruct2>;
+
+template <typename T>
+using TemplateProvidedArgumentNotUsed = TplDirectStruct7<IndirectStruct2, T>;
+
+template <typename T>
+using TemplateNotProvidedArgumentNotUsed =
+    TplIndirectStruct3<IndirectStruct2, T>;
+
 /**** IWYU_SUMMARY
 
 tests/cxx/iwyu_stricter_than_cpp-type_alias.h should add these lines:

--- a/tests/cxx/iwyu_stricter_than_cpp.cc
+++ b/tests/cxx/iwyu_stricter_than_cpp.cc
@@ -274,6 +274,11 @@ void TestAliasTemplates() {
   // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
   // IWYU: TplIndirectStruct3 is...*iwyu_stricter_than_cpp-i5.h
   TemplateNotProvidedArgumentNotUsed<IndirectStruct3> tnpanu;
+
+  TemplateProvidedArgumentUsed<DoesNotForwardDeclareAndIncludes>
+      argument_provided;
+  // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
+  TemplateProvidedArgumentUsed<DoesEverythingRight> argument_not_provided;
 }
 
 void TestAutocast() {

--- a/tests/cxx/iwyu_stricter_than_cpp.cc
+++ b/tests/cxx/iwyu_stricter_than_cpp.cc
@@ -259,6 +259,21 @@ void TestAliasTemplates() {
   DoesEverythingRightAlTpl<int> dor(4);
   // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
   (void)&DoesEverythingRightAlTpl<int>::a;
+
+  // IWYU: IndirectStruct3 needs a declaration
+  // IWYU: IndirectStruct3 is...*iwyu_stricter_than_cpp-i3.h
+  TemplateProvidedArgumentUsed<IndirectStruct3> tpau;
+  // IWYU: IndirectStruct3 needs a declaration
+  // IWYU: IndirectStruct3 is...*iwyu_stricter_than_cpp-i3.h
+  // IWYU: TplIndirectStruct3 is...*iwyu_stricter_than_cpp-i5.h
+  TemplateNotProvidedArgumentUsed<IndirectStruct3> tnpau;
+  // IWYU: IndirectStruct3 needs a declaration
+  // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
+  TemplateProvidedArgumentNotUsed<IndirectStruct3> tpanu;
+  // IWYU: IndirectStruct3 needs a declaration
+  // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
+  // IWYU: TplIndirectStruct3 is...*iwyu_stricter_than_cpp-i5.h
+  TemplateNotProvidedArgumentNotUsed<IndirectStruct3> tnpanu;
 }
 
 void TestAutocast() {
@@ -479,7 +494,7 @@ The full include-list for tests/cxx/iwyu_stricter_than_cpp.cc:
 #include "tests/cxx/iwyu_stricter_than_cpp-i3.h"  // for IndirectStruct3
 #include "tests/cxx/iwyu_stricter_than_cpp-i4.h"  // for IndirectStruct4
 #include "tests/cxx/iwyu_stricter_than_cpp-i5.h"  // for TplIndirectStruct3
-#include "tests/cxx/iwyu_stricter_than_cpp-type_alias.h"  // for DoesEverythingRightAl, DoesEverythingRightAlTpl, DoesNotForwardDeclareAl, DoesNotForwardDeclareAlTpl, DoesNotForwardDeclareAndIncludesAl, DoesNotForwardDeclareAndIncludesAlTpl, DoesNotForwardDeclareProperlyAl, IncludesAl, IncludesAlTpl, IncludesElaboratedAl, IndirectStruct3NonProvidingAl, IndirectStruct4NonProvidingAl, TplAllForwardDeclaredAl, TplAllNeededTypesProvidedAl, TplDoesEverythingRightAgainAl, TplDoesEverythingRightAl, TplDoesNotForwardDeclareAl, TplDoesNotForwardDeclareAndIncludesAl, TplDoesNotForwardDeclareProperlyAl, TplIncludesAl, TplOnlyArgumentTypeProvidedAl, TplOnlyTemplateProvidedAl
+#include "tests/cxx/iwyu_stricter_than_cpp-type_alias.h"  // for DoesEverythingRightAl, DoesEverythingRightAlTpl, DoesNotForwardDeclareAl, DoesNotForwardDeclareAlTpl, DoesNotForwardDeclareAndIncludesAl, DoesNotForwardDeclareAndIncludesAlTpl, DoesNotForwardDeclareProperlyAl, IncludesAl, IncludesAlTpl, IncludesElaboratedAl, IndirectStruct3NonProvidingAl, IndirectStruct4NonProvidingAl, TemplateNotProvidedArgumentNotUsed, TemplateNotProvidedArgumentUsed, TemplateProvidedArgumentNotUsed, TemplateProvidedArgumentUsed, TplAllForwardDeclaredAl, TplAllNeededTypesProvidedAl, TplDoesEverythingRightAgainAl, TplDoesEverythingRightAl, TplDoesNotForwardDeclareAl, TplDoesNotForwardDeclareAndIncludesAl, TplDoesNotForwardDeclareProperlyAl, TplIncludesAl, TplOnlyArgumentTypeProvidedAl, TplOnlyTemplateProvidedAl
 #include "tests/cxx/iwyu_stricter_than_cpp-typedefs.h"  // for DoesEverythingRight, DoesNotForwardDeclare, DoesNotForwardDeclareAndIncludes, DoesNotForwardDeclareProperly, Includes, IncludesElaborated, IndirectStruct3NonProvidingTypedef, IndirectStruct4NonProvidingTypedef, TplAllForwardDeclared, TplAllNeededTypesProvided, TplDoesEverythingRight, TplDoesEverythingRightAgain, TplDoesNotForwardDeclare, TplDoesNotForwardDeclareAndIncludes, TplDoesNotForwardDeclareProperly, TplIncludes, TplOnlyArgumentTypeProvided, TplOnlyTemplateProvided
 struct DirectStruct1;
 struct DirectStruct2;

--- a/tests/cxx/iwyu_stricter_than_cpp.cc
+++ b/tests/cxx/iwyu_stricter_than_cpp.cc
@@ -251,6 +251,16 @@ void TestTypeAliases() {
   WithAlias<int, TplIndirectStruct2<int>> wa;
 }
 
+void TestAliasTemplates() {
+  DoesNotForwardDeclareAlTpl<int> dnfd(1);
+  IncludesAlTpl<int> i(2);
+  DoesNotForwardDeclareAndIncludesAlTpl<int> dnfdai(3);
+  // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
+  DoesEverythingRightAlTpl<int> dor(4);
+  // IWYU: IndirectStruct2 is...*iwyu_stricter_than_cpp-i2.h
+  (void)&DoesEverythingRightAlTpl<int>::a;
+}
+
 void TestAutocast() {
   // We need full type of IndirectStruct2 because the declarer of the following
   // functions didn't.
@@ -469,7 +479,7 @@ The full include-list for tests/cxx/iwyu_stricter_than_cpp.cc:
 #include "tests/cxx/iwyu_stricter_than_cpp-i3.h"  // for IndirectStruct3
 #include "tests/cxx/iwyu_stricter_than_cpp-i4.h"  // for IndirectStruct4
 #include "tests/cxx/iwyu_stricter_than_cpp-i5.h"  // for TplIndirectStruct3
-#include "tests/cxx/iwyu_stricter_than_cpp-type_alias.h"  // for DoesEverythingRightAl, DoesNotForwardDeclareAl, DoesNotForwardDeclareAndIncludesAl, DoesNotForwardDeclareProperlyAl, IncludesAl, IncludesElaboratedAl, IndirectStruct3NonProvidingAl, IndirectStruct4NonProvidingAl, TplAllForwardDeclaredAl, TplAllNeededTypesProvidedAl, TplDoesEverythingRightAgainAl, TplDoesEverythingRightAl, TplDoesNotForwardDeclareAl, TplDoesNotForwardDeclareAndIncludesAl, TplDoesNotForwardDeclareProperlyAl, TplIncludesAl, TplOnlyArgumentTypeProvidedAl, TplOnlyTemplateProvidedAl
+#include "tests/cxx/iwyu_stricter_than_cpp-type_alias.h"  // for DoesEverythingRightAl, DoesEverythingRightAlTpl, DoesNotForwardDeclareAl, DoesNotForwardDeclareAlTpl, DoesNotForwardDeclareAndIncludesAl, DoesNotForwardDeclareAndIncludesAlTpl, DoesNotForwardDeclareProperlyAl, IncludesAl, IncludesAlTpl, IncludesElaboratedAl, IndirectStruct3NonProvidingAl, IndirectStruct4NonProvidingAl, TplAllForwardDeclaredAl, TplAllNeededTypesProvidedAl, TplDoesEverythingRightAgainAl, TplDoesEverythingRightAl, TplDoesNotForwardDeclareAl, TplDoesNotForwardDeclareAndIncludesAl, TplDoesNotForwardDeclareProperlyAl, TplIncludesAl, TplOnlyArgumentTypeProvidedAl, TplOnlyTemplateProvidedAl
 #include "tests/cxx/iwyu_stricter_than_cpp-typedefs.h"  // for DoesEverythingRight, DoesNotForwardDeclare, DoesNotForwardDeclareAndIncludes, DoesNotForwardDeclareProperly, Includes, IncludesElaborated, IndirectStruct3NonProvidingTypedef, IndirectStruct4NonProvidingTypedef, TplAllForwardDeclared, TplAllNeededTypesProvided, TplDoesEverythingRight, TplDoesEverythingRightAgain, TplDoesNotForwardDeclare, TplDoesNotForwardDeclareAndIncludes, TplDoesNotForwardDeclareProperly, TplIncludes, TplOnlyArgumentTypeProvided, TplOnlyTemplateProvided
 struct DirectStruct1;
 struct DirectStruct2;

--- a/tests/cxx/typedef_in_template.cc
+++ b/tests/cxx/typedef_in_template.cc
@@ -122,6 +122,9 @@ struct Identity {
   struct Inner {
     using Type = T;
   };
+
+  static T t;
+  using SugaredType = decltype(t);
 };
 
 template <typename T>
@@ -168,6 +171,10 @@ void ArgumentTypeProvision() {
   // IWYU: IndirectClass needs a declaration
   Outer<IndirectClass*>::Inner<Providing>::AliasedTpl ptr_p;
 }
+
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+constexpr auto s3 = sizeof(Identity<IndirectClass>::SugaredType);
 
 /**** IWYU_SUMMARY
 

--- a/tests/cxx/typedef_in_template.cc
+++ b/tests/cxx/typedef_in_template.cc
@@ -125,6 +125,9 @@ struct Identity {
 
   static T t;
   using SugaredType = decltype(t);
+
+  template <int>
+  using AliasTemplate = T;
 };
 
 template <typename T>
@@ -170,6 +173,15 @@ void ArgumentTypeProvision() {
   Outer<Providing>::Inner<IndirectClass*>::AliasedTpl p_ptr;
   // IWYU: IndirectClass needs a declaration
   Outer<IndirectClass*>::Inner<Providing>::AliasedTpl ptr_p;
+
+  Identity<Providing>::AliasTemplate<1> atp;
+  (void)sizeof(Identity<Providing>::AliasTemplate<1>);
+  // IWYU: NonProviding is...*typedef_in_template-i1.h
+  // IWYU: IndirectClass is...*indirect.h
+  Identity<NonProviding>::AliasTemplate<1> atn;
+  // IWYU: NonProviding is...*typedef_in_template-i1.h
+  // IWYU: IndirectClass is...*indirect.h
+  (void)sizeof(Identity<NonProviding>::AliasTemplate<1>);
 }
 
 // IWYU: IndirectClass needs a declaration

